### ISSUE TITLE
Standardize branching to match OpenSearch

### ISSUE THREAD

**Issue by dblock** [opensearch-project/index-management#81]
Coming from [.github#13](https://github.com/opensearch-project/.github/issues/13), standardize release branching to match what [OpenSearch](https://github.com/opensearch-project/OpenSearch) is [doing](https://github.com/opensearch-project/.github/blob/main/RELEASING.md#branching).

1. Create a `1.0` branch for the OpenSearch 1.0 release, `1.x` branch for next 1.1 release, and use `main` for 2.0 development. Make sure CI is enabled on those.
2. Update your release/branching documentation. If you don't have one, add a RELEASING.md that links to, or has content from [.github/RELEASING.md](https://github.com/opensearch-project/.github/blob/main/RELEASING.md).
3. If you are using a `develop` branch, stop. You should be using the default `main` brach for furthermost, future development. 
4. Delete any merged/stale/develop/feature branches that are no longer in use.
5. Communicate any changes in process to your team.

Refer to [release branching](https://github.com/opensearch-project/.github/blob/main/RELEASING.md#branching) for more information.

**Comment by thalurur** [comment]
Hi @dblock 

```
Given the current major release of 1.0, projects in this organization maintain the following active branches.

    main: The next major release. This is the branch where all merges take place and code moves fast.
    1.x: The next minor release. Once a change is merged into main, decide whether to backport it to 1.x.
    1.0: The current release. In between minor releases, only hotfixes (e.g. security) are backported to 1.0.

```

I am a little confused with this ask:
* From reading above I get every commit/PR goes to `main`  is that correct? Someone will need to port these to `1.0` or `1.x` if they need to be - who is going to determine that these need to be ported over and do these (the contributor for original commit or a maintainer)? and is there going to be a new PR to backport? (meaning direct push to these branches is blocked)
* Are we going to have a new `n.0` branch and `n.x` branch after a new major version `n` is launched and similarly a new `n.y` branch for each new `y` minor release for `n` major version? or these branches (1.0/1.x) will be reset after each major/minor version launch?

**Comment by dblock** [comment]
> I am a little confused with this ask:
> 
> * From reading above I get every commit/PR need to go to `main` first regardless, is that correct? Someone will need to port these to `1.0` or `1.x` if they need to be - who is going to determine that these need to be ported over and do these (the contributor for original commit or a maintainer)? and is there going to be a new PR to backport? (meaning direct push to these branches is blocked)

Correct. Maintainers should determine what should be backported in these branches. 

I think we should add text about backporting to https://github.com/opensearch-project/.github/blob/main/RELEASING.md -  want to take a stab at it?

> * Are we going to have a new `n.0` branch and `n.x` branch after a new major version `n` is launched and similarly a new `n.y` branch for each new `y` minor release for `n` major version? or these branches (1.0/1.x) will be reset after each major/minor version launch?

The former. If someone gets tired of seeing a 1.x branch and we are done making 1.x releases, we can delete the branch (since we tag actual releases, theoretically you can re-create a branch based on the tag).

**Comment by thalurur** [comment]
@dblock Thanks for the clarification.

Can you also shed some light on why this is a requirement for all the plugins? I checked with the team and this doesn't seem very intuitive to do by all plugins as this brings in additional overhead and not immediately seeing value. It will help use if you can shed light on what this will solve and why this is a requirement.

**Comment by CEHENKLE** [comment]
Hi Ravi --
As part of moving to OpenSearch, we're trying to standardize a set of practices across the organization, not just at the individual repo level.  There are several reason for this:

1/ We want to make it easy as possible for new developers to commit to OpenSearch.  If every plugin functions slightly differently, it adds friction to that process.   As an aside, this is the same branching strategy that's used by Lucene, so this brings us into a more standard pattern for Open Source projects :)

2/ We are working on new tooling for our build systems.  We need to get to the point where we have CI/CD running to build the entire artifact (not just an individual repo) so we can run things like performance testing every night, ideally on the main, 1.0 and 1.x branches.  Again, this is much more difficult if we're trying to accommodate different branching patterns.

Thanks,
/C
